### ISSUE TITLE
MediaSource in Worker

### ISSUE THREAD

**Issue by DanielBaulig** [w3c/media-source#175]
The Videos team at Facebook has a growing need to move calls to Media Source Extensions off of the main thread and into a Worker. While most parts of your off-the-shelve streaming video player can be implemented in a Worker (fetching and parsing of the manifest as well as selecting and fetching of segments) the calls to and interactions with the Media Source Extensions API (e.g. `SourceBuffer.prototype.appendBuffer`, `SourceBuffer.prototype.remove`, etc) remain a bottleneck and have to happen on the main browser rendering thread. 

On a sizable and complex web site large blocks of JS execution can regularly block the main thread making videos slow to start playing and causing lots of rebuffering, especially in live videos where you can only buffer ahead very little data. To ensure great web video performance in the long run, even as web apps continue to grow in size and complexity, we believe we need to allow video players to move into Workers and off of the main thread.

We would like to propose this change to the working group and involve other API users as well as implementors in the discussion.

**Comment by bmaurer** [comment]
#14 also seems relevant here -- based on my reading of that bug it would help with part of this problem (once you fetch a segment you'd pass a stream to that audio to MediaSource and would bypass the main thread, or even a worker thread in terms of getting that data to where it needs to go).

But that still leaves the issue of fetching a manifest and initiating a request for a segment

**Comment by jyavenard** [comment]
> the calls to and interactions with the Media Source Extensions API (e.g. SourceBuffer.prototype.appendBuffer, SourceBuffer.prototype.remove, etc) remain a bottleneck and have to happen on the main browser rendering thread.

That it's a bottleneck would be mostly a user agent implementation issue.
Those APIs are designed to be asynchronous, they should take no time at all and not block the main thread.
With current Firefox, running those two operations in a worker would make little (or no) performance difference, they are in effect running internally in a worker.

I wish issue #100 progressed further, it would certainly facilitate handling those operations.

**Comment by bmaurer** [comment]
The issue here isn't the cost of video *on* the main thread. The issue is that if the main thread is busy with other work (say the javascript needed to render a page) that it can not be responsive to input from the network. This means that the video will be slower to buffer even though the client has been able to get the video over the network.

The goal is for the video loading code to be isolated on a worker thread where it can operate in real time without the risk of latency induced by work on the main thread.

**Comment by zhaoz** [comment]
We are seeing the same issue on YouTube and I'd like to add our voice to support a solution here as well. Moving append buffer off the main thread would greatly simplify the juggling we need to do to prioritize video playback.

**Comment by bmaurer** [comment]
It's great to see that youtube has the same take on this issue. Would it be worth taking the next step here and drafting out what a spec change would look like? I think we can model this on the offscreen canvas api (https://hacks.mozilla.org/2016/01/webgl-off-the-main-thread/)

> var canvas = document.getElementById('myCanvas');
> var offscreen = canvas.transferControlToOffscreen();
> var worker = new Worker('worker.js');
> worker.postMessage({ canvas: offscreen }, [offscreen]);

Essentially MediaSource would offer an API which would pass control of the media source APIs to a worker thread. Any APIs that MediaSource offers (eg to add source buffers) would only be callable from the worker.

MediaSource is a fairly complex API with a number of events / methods. It'd be great to get some feedback from @zhaoz, @DanielBaulig as well as folks who work on browsers and this spec to comment on what kind of threading issues might come up with this API.

**Comment by DanielBaulig** [comment]
The main limitation with transferring MediaSource object that I can currently see is that in general most video player implementations heavily depend on certain video node attribute for determining which video segments to fetch and play. These are as far as I can tell primarily `video.currentTime`, `video.paused` and `video.buffered`. 
The main thread will need to regularly update the worker with the current values of these for most video player implementations to be able to operate correctly. 

However, the API itself could have support for this: the worker could get copies of these values on every event loop exposed. This exposure could happen as part of the transferred MediaSource object, e.g. on a bag of exposed values or directly on the object:

    onmessage = function(message) {
      if (message.mediaSource.mediaElementState.paused) {
        // don't do anything while the video is paused
        return;
      }
    };

Another problem may be that the main thread indirectly retains a reference to the MediaSource as an object URL on the videos `src` attribute. I am not sure how this will complicate the safe usage of object URLs.

**Comment by bmaurer** [comment]
That makes total sense. It does seem like the MediaSource API should be responsible for relaying that information -- accessing it directly from the video element seems like it violates layering and would make the threading far more difficult here.

One other thing we should think about -- maybe this is a totally separate issue -- I worry about the latency of starting the worker, downloading the necessary JS, etc. On Facebook we've found that the amount of JS necessary to have the logic for streaming the video can be pretty high. I wonder if there's a way to allow the server to list the URLs necessary to start streaming so that they can be fetched and inserted into the buffer while the more advanced control starts up.

**Comment by DanielBaulig** [comment]
> One other thing we should think about -- maybe this is a totally separate issue -- I worry about the latency of starting the worker, downloading the necessary JS, etc. On Facebook we've found that the amount of JS necessary to have the logic for streaming the video can be pretty high. I wonder if there's a way to allow the server to list the URLs necessary to start streaming so that they can be fetched and inserted into the buffer while the more advanced control starts up.

I don't think there's anything preventing video player implementations from doing something like this currently. That said, in the context of a transferable MediaSource, this becomes a little harder. This could potentially mean creating a MediaSource object and several sourceBuffers on the main thread, appending data to the sourceBuffers and only later transferring ownership to the Worker. This transfer won't be simple anymore though since there's a complex mix of objects and possibly tasks that has to be dealt with.

**Comment by jyavenard** [comment]
One of the issue in dealing with a sourcebuffer in a worker thread, is how, currently, the source buffer notifies that the current operation has completed: through events fired on the attached media element and which is only ever accessed on the main thread.

So if your main thread is busy, no update/updateend event and you're back to the issue of not being able to keep up with your source buffer.

This problem could be easily worked around if rather than relying on events such as update/updateend to know when you can feed new data to the source buffer we used Promises.
I opened an issue a while back to implement this:
https://github.com/w3c/media-source/issues/100

The beauty with SourceBuffer::appendBuffer and SourceBuffer::remove with returning Promises is that it would be fully backward compatible. Once the appendBuffer operation has completed, resolve the Promise *and* queue update/updateend
Up to the player to use one mechanism or the other, never both.

Making a SourceBuffer usable on a worker thread wouldn't be a too complicated task (famous last words) in gecko, the architecture is already asynchronous and thread-safe.
The interaction with the media element would be the issue.

But if we limit the interaction to the source buffer to appendBuffer and remove operations, and checking of the buffered range to be performed on the source buffers alone, and we use Promises then it should be straight forward.

**Comment by wolenetz** [comment]
@https://github.com/w3c/media-source/issues/175#issuecomment-278253566, 'updateend' is fired on the SourceBuffer , not HTMLMediaElement. However, readyState transitions (and consequent events like 'loadedmetadata') *would* occur on the media element on the main thread.

I'm less clear that Promises are required for MSE-on-SW, though I'm not opposed (I just don't want to tangle the two features).

**Comment by wolenetz** [comment]
Here's a summary of this issue so far, from my understanding. I have a set of questions, below, for which web author input would be greatly appreciated.

Proposals in [2] and [3] help to limit the scope of this to be specific to just MSE, and not require exposing all of HTMLMediaElement API also to Service Worker (SW):
 
[2] Proposes that the MediaSource API be made “transferrable” to a SW
[3] Adds that major HTMLMediaElement attributes necessary for MSE API usage be periodically copied from the main thread to SW (perhaps as part of the transferred MediaSource object).
 
[1] https://github.com/w3c/media-source/issues/175
[2] https://github.com/w3c/media-source/issues/175#issuecomment-271909355
[3] https://github.com/w3c/media-source/issues/175#issuecomment-272210121

Questions:
* Should the “transfer” be allowed on a MediaSource object which has already had SourceBuffers added to it? (This might allow for initial buffering to occur on main thread on page load, while the SW context is still spinning up.) (I’m not clear this will help so much in practice, as some web players already experience high time to first frame due to other activities contending on the main thread immediately on page load.)
* Is there potential “safe usage of Object URLs” problem with transferring the MediaSource context to a SW? (I would defer to web platform security experts. I suspect that a set of reasonable restrictions/checks on the transfer may be needed. Any suggestions?)
* What portion of the MediaSource and SourceBuffer API need to be transferred?
* What portion of the HTMLMediaElement state need to be periodically copied into the transferred MediaSource’s SW context?
* Precisely when should such state be copied?
* Are any HTMLMediaElement events needed in the SW context? (This would increase complexity; I’m not certain any are needed in SW context, since the web app could forward these -- though with consequent potentially high latency -- into the SW context itself.)
* Once transferred to SW, would the original MediaSource be restricted from access from any other context?
* I assume the HTMLMediaElement itself must provide continued access; though would queries of it like “buffered”, which MSE extends, need to round-trip into the SW context?

**Comment by DanielBaulig** [comment]
Am I assuming correctly that you are using the term Service Worker and the general Worker API interchangeably? 

I'll try to answer most of the questions for which I believe I have enough context:

> Should the “transfer” be allowed on a MediaSource object which has already had SourceBuffers added to it? (This might allow for initial buffering to occur on main thread on page load, while the SW context is still spinning up.) (I’m not clear this will help so much in practice, as some web players already experience high time to first frame due to other activities contending on the main thread immediately on page load.)

This is probably not necessary. The `sourceopen` event takes some time to fire and the application probably needs time to load data to append to the source buffers too. If those steps take significantly less time than spinning up the Worker than there may be some room for optimization, but I believe that this is likely better done by improving Worker startup time, than introducing bloat here.

> Is there potential “safe usage of Object URLs” problem with transferring the MediaSource context to a SW? (I would defer to web platform security experts. I suspect that a set of reasonable restrictions/checks on the transfer may be needed. Any suggestions?)

I am not sure I can answer this question.

> What portion of the MediaSource and SourceBuffer API need to be transferred?

We've spent some time internally thinking about this and we came to the conclusion that the entirety of the SourceBuffer API is probably the minimally viable subset of the API for the application.

Having the MediaSource object itself available is preferred but not *necessarily* required. We believe though due to the amount of dependencies between the MediaSource object and it's SourceBuffer objects, that transferring the MediaSource object is likely the easier approach from a spec and implementation perspective.

We'd love to hear some thoughts from implementors on this though.

> What portion of the HTMLMediaElement state need to be periodically copied into the transferred MediaSource’s SW context?

> Are any HTMLMediaElement events needed in the SW context? (This would increase complexity; I’m not certain any are needed in SW context, since the web app could forward these -- though with consequent potentially high latency -- into the SW context itself.)

The Worker would generally require access to information that is helpful in making decisions on how much, how fast and what kind of data to buffer. Generally this could be the current playhead (currentTime), the current state of the video buffer (buffered attribute) and current payback rate (playbackRate), if audio is audible (muted and volume) and probably more.
The exact information here however is highly dependent on the details and sophistication of the application and probably highly vary from application to application. Note also, that a small delay in updating the Worker with this information should not lead to any significant problems for the application. In fact, there are plenty of applications and scenarios conceivable in which no state synchronization at all would be required. I think generally the same is true for any HTMLMediaElement events.

For those reasons and to make the specification and implementation process easier we would probably suggest to leave this synchronization completely up to the application. The application can use postMessage to update the any necessary state as required.

> Once transferred to SW, would the original MediaSource be restricted from access from any other context?

I think that is the current understanding.

> I assume the HTMLMediaElement itself must provide continued access; though would queries of it like “buffered”, which MSE extends, need to round-trip into the SW context?

This seems to be more of an implementation / specification related question that I am not entirely sure I am able to give feedback on. My understanding here however is that the buffered attribute both for sourceBuffer as well as the video node already need information from the (already) asynchronous and in a different thread running process of media decoding.

**Comment by wolenetz** [comment]
Thanks for you answers - those do help describe the possible feature scope more precisely. w.r.t.:

> Am I assuming correctly that you are using the term Service Worker and the general Worker API interchangeably?

You're indeed correct - I had mistakenly constrained the scope too far :). Unless the general Worker API is too broad, let's target it, not just Service Workers.

**Comment by wolenetz** [comment]
@https://github.com/w3c/media-source/issues/175#issuecomment-299345346:
>  My understanding here however is that the buffered attribute both for sourceBuffer as well as the video node already need information from the (already) asynchronous and in a different thread running process of media decoding.

Decoding itself doesn't impact buffered ranges (unless there's a decode error and transition of MediaSource readyState to closed.) Rather, operations on the MSE API itself can change the buffered ranges. What I was trying to get at is: what are the potentially required synchronization points between the HTMLMediaElement attached by a transferred MSE worker context? For example:

* HTMLMediaElement.buffered and duration: If an app synchronously queries the 'buffered' attribute of such an HTMLMediaElement, yet the MSE in worker context is in the process of, or perhaps has recently completed, a SourceBuffer appendBuffer(), SourceBuffer remove(),  SourceBuffer mode or timestampOffset change when previously 'ended', MediaSource endOfStream(), or MediaSource set duration operation, is there some guarantee of synchronization of the result of such operation that needs to be handled? What should such a guarantee look like? Note that actual playback/decode doesn't factor into this (buffered time range updates due to appendBuffer()/remove()/duration setting don't require actual decode). I assume there needs to be some specified synchronization of such results; in non-worker/current case, if the SourceBuffer.updating attribute is false across all SourceBuffers in the attached MediaSource, then the buffered range query on the parent media element should be stable and correct. Is some similar guarantee required in the worker case? 

* HTMLMediaElement.seekable and set/clearLiveSeekableRanges (also involves HTMLMediaElement.buffered state): ditto: should there be some guarantee of synchronization with the worker before a query of seekable can be satisfied?

* Worker MediaSource.buffered (potential new API method): Note that the worker context can itself inspect and intersect manually the SourceBuffer buffered ranges to mimic what the parent element's logic would do. Should the transferred MSE API provide such copy of the parent media element's logic that can be processed without synchronizing state with the parent media element?

**Comment by bmaurer** [comment]
These are great questions. I think my intuition is that generally the worker thread which holds the MSE will be communicating with the main thread, eg via postMessage. I think it would make sense to enforce logical causality. For example, if you call appendBuffer, then send a message to the main thread the result of that appendBuffer should be observable when processing the send message.

I agree that MediaSource.buffered probably makes sense -- there's no need to query the main thread for this information.

**Comment by wolenetz** [comment]
> For example, if you call appendBuffer, then send a message to the main thread the result of that appendBuffer should be observable when processing the send message.

To be clear, I assume you mean more like "if you call appendBuffer, and upon later receiving updateend you send a message to the main thread, the result of the appendBuffer should be visible on the main thread when processing the receipt of that message." Please correct me if I misunderstand :)

**Comment by wolenetz** [comment]
Also, does it make sense to use such a transferred MSE API from a Service Worker context (I don't think so, because such a context is independent of a web page, and therefore from a DOM containing a media element). ISTM, this makes sense from a web worker context, not from SW context.

**Comment by bmaurer** [comment]
> To be clear, I assume you mean more like "if you call appendBuffer, and upon later receiving updateend you send a message to the main thread, the result of the appendBuffer should be visible on the main thread when processing the receipt of that message." Please correct me if I misunderstand :)

yep, exactly.

> Also, does it make sense to use such a transferred MSE API from a Service Worker context (I don't think so, because such a context is independent of a web page, and therefore from a DOM containing a media element). ISTM, this makes sense from a web worker context, not from SW context.

By worker I meant web worker here

**Comment by DanielBaulig** [comment]
I agree with what @bmaurer said: maintaining logical causality with postMessage probably makes sense.

If we are considering replicating the `buffered` HTMLMediaElement extension on MediaSource - which I totally think makes sense -, we should probably do the same thing for `seekable`.

**Comment by wolenetz** [comment]
Thanks for your responses to my questions. Here's my newbie-to-workers question: can a Worker do a fetch that is intercepted by a Service Worker, thereby enabling a combination of offline MSE-in-Worker being fed by data provided by the Service Worker? Where are known gaps in spec or implementation to accomplish such a scenario (other than, of course, this issue's MSE-in-Worker discussion :) )?

**Comment by kentuckyfriedtakahe** [comment]
I'm with @jyavenard thinking that adding #100 while workerising SourceBuffer would be worthwhile. It may be timely because the worker codepath may end up being a fork of the non-worker path. I'd like to hear from @bmaurer and @zhaoz about whether this will make the player implementation simpler or better.

**Comment by bakulf** [comment]
I would like to share my point of view about the transferring of MediaSource and SourceBuffer objects to workers. If I have understood the spec correctly, a normal use of a MediaSource is via `HTMLMediaElement.src = URL.createObjectURL(MS)`.  The output of URL.createObjectURL() is a DOMString, fully cloneable via postMessage().

Would be probably cleaner to expose MediaSource and SourceBuffer constructors to workers.
A worker can create a MediaSource and SourceBuffer(s) then, simply send a blob URL to the window: `worker.onmessage = e => { mediaElement.src = e.data; }`

The change I propose is mark MediaSource interface in this way:

```
[Constructor, Exposed=(Window,Worker)]
interface MediaSource {...
```
same for SourceBuffer.

Note that SourceBuffer supports the appending of ArrayBuffer and ArrayBufferView. Both are already transferable to workers.
Another important point, maybe the most important, so far, EventTargets are not transferable to workers because EventHandler cannot be handled correctly when the object is transferred.

**Comment by alastor0325** [comment]
Agree with [comment](https://github.com/w3c/media-source/issues/175#issuecomment-327715837), it would be easy if we restrict the media source and source buffer could only be used in one thread all the time and disallow them being transferrable.

If they’re transferable, how should the behavior of event handlers be defined? It seems a difficult problem, and the [spec](https://html.spec.whatwg.org/multipage/structured-data.html#transferable-objects) doesn't mention anything about it. 

eg. 
If I transfer an media source before its “onsourceopen” be called, which thread the event should be dispatched?
If the source buffer can also be transferred, whether we can transfer multiple source buffers to multiple workers?
…e.t.c

—

If what I understand it correctly, the usage scenario would like the following.

First, we create both media source and source buffer on the WT, and send the URL back to MT after wrapping media source as URL. After the media source is attached successfully to media element, “sourceopen” would be triggered on WT. Next, MT should send the video URL to WT, then all things could be done independently inside the WT from fetching data to appending data into buffer (and with some new API for getting the status of media element).

Please correct me if I was wrong and welcome to propose other possible scenarios.

**Comment by DanielBaulig** [comment]
I see the intentions here and agree that this would make several aspects easier. The only concern I have is that this would add a significant amount of additional startup time.

On top of the time it takes to complete `sourceopen`, now we may also have to wait for a Worker to spin up and a message to be sent form the Worker to the main thread (which may be busy doing other stuff). All these things will have to happen in sequence and it will require two dedicated time slices on the main thread to complete.

In a scenario with a transferable MediaSource, waiting for `sourceopen` and the bootstrapping of the Worker can happen in parallel and the Worker will not have to round trip back to the main thread to complete the transaction reducing the main thread time slices require to one.

The biggest risks associated with moving MSE into Workers from a product perspective is probably the startup latency. The solution as suggested in https://github.com/w3c/media-source/issues/175#issuecomment-327715837 increases this risk significantly.

**Comment by alastor0325** [comment]
This might be a stupid question, why couldn't we bootstrap worker in advance, before starting playing video?

Can we bootstrap worker in the beginning? and then use `postMessage()` to notify worker to return the URL of media source when we would start MSE soon?

> In a scenario with a transferable MediaSource, waiting for sourceopen and the bootstrapping of the Worker can happen in parallel and the Worker will not have to round trip back to the main thread to complete the transaction reducing the main thread time slices require to one. 

If we can start worker in advance, the following might be the possible use case. In this case, MT only need one time slice to handle the message sent from WT.

```javascript
worker.postMessage(url /* used for fetching on WT */);
worker.onMessage = e => { mediaElement.src = e.data; };
```

**Comment by alastor0325** [comment]
The main problem to make media source transferable is "how to handle the event handler of transferable object", because it's undefined behavior.

I think this topic is out of the media scope, and I would like to know how strong intention we want to go with this direction? If do so, we need to discuss and modify with worker's working group and make sure the behavior can be well defined.

However, it might be take a long time, and @bakulf, who is our worker expert, also suggest don't try to transfer an object with event handler.

**Comment by DanielBaulig** [comment]
@alastor0325 We can only bootstrap the Worker ahead of time, if we have an opportunity to do so. If the first thing we are doing on a blank page load is to create and assign a MediaSource, there's no way for us to bootstrap a Worker ahead of time. 

In addition to that, creating the object URL on the Worker thread will require us to touch the main thread twice. Consider the following example:

```javascript
    /* main thread code */
    // first main thread execution
    const video = document.querySelector('video');
    const worker = new Worker('mse_worker.js');
    worker.addEventListener('message', (event) => {
      // second main thread execution
      video.src = event.data;
    });
```

But if we can create the MediaSource on the main thread, we only need one time slice on the main thread:

```javascript
    /* main thread code */
    // first main thread execution
    const video = document.querySelector('video');
    const worker = new Worker('mse_worker.js');
    const ms = new MediaSource();
    const url = window.URL.createObjectURL(ms);
    video.src = url;
    worker.postMessage(ms, [ ms ]);
```

We don't need a second time slice on the main thread before we can start loading video.

Talking to some folks at Mozilla, they wanted to look into why the `sourceopen` event even is required/asynchronous in the first place.

Another option could be to require transferring of the MediaSource to the Worker *before* the URL assignment happens. In that case any events firing on the MediaSource could be deterministically scheduled to execute on the Worker instead of the main thread.

Lastly, I could imagine something TransferableMediaSource, that internally does all the steps MediaSource would do upon creation, but doesn't expose any external API interface except for a getMediaSource() method. After the method was called, the TransferableMediaSource looses it's ability to be transferred.
This could be created on the main thread, assigned to the video node's src through an Object URL and then transferred to the Worker, where the Worker would call getMediaSource to obtain a reference on the "true", underlying MediaSource.

**Comment by jyavenard** [comment]
Why not only having the source buffers in a worker?
They really are the only ones we care about when it comes to feeding data fast enough.

**Comment by alastor0325** [comment]
@DanielBaulig I think it's a good idea to have "TransferableMediaSource", it could fix the problem about the event handler and also it would only need one time slice on the main thread.

In addition, you've mention we would need something like "mediaSource.mediaElementState" which can be reflected the status of media element. Maybe we could start to define the specific attributes which we need.

**Comment by DanielBaulig** [comment]
> In addition, you've mention we would need something like "mediaSource.mediaElementState" which can be reflected the status of media element. Maybe we could start to define the specific attributes which we need.

For the most part I think it's sufficient to provide these media element states via postMessage to the worker at a regular rate. 

Unless the platform can significantly improve update performance or maybe even provide thread-safe, but otherwise real-time variations of these, I don't think they are of highest priority.

The most important states/properties are probably `currentTime` and `error`. Virtually any media related property can be useful though to optimize certain conditions, prevent over- or underfetching and allow for a more consistent player and data fetching state (e.g. `buffered`, `playbackRate`, `paused`, `seeking`, `ended`, `readyState`, `duration`).

> Why not only having the source buffers in a worker?
> They really are the only ones we care about when it comes to feeding data fast enough.

Generally this is correct. I remember that this was brought up before as an option, but I'm not entirely sure what the arguments against it were. There's definitely the limitation of having to touch the main thread twice though (you have to wait for `'sourceopen'` to fire before you can create SourceBuffer instances), but I believe the major concern was about how to model the APIs around this, since SourceBuffers are exposed on several MediaSource APIs and all of them would become invalid when the SourceBuffer is transferred into a Worker.

**Comment by wolenetz** [comment]
I'm considering putting together a proposal to incubate MSE-in-Workers in WICG (similar to how we've been incubating codec switching). I've read through the history on this issue so far and am curious if there are any new concerns, suggestions or use cases folks might have since ~1 year ago.

**Comment by estobbart** [comment]
To me this is sort of at odds with Google's policy on autoplay & the [MEI](https://developers.google.com/web/updates/2017/09/autoplay-policy-changes#mei).
If video is being done in a worker, it would hint to me that video playback isn't the primary objective of the user on the site. If it is the primary objective, can't the other rendering operations be limited instead of tying up the main loop?
I do see some advantages in preparing MediaSource/SourceBuffers in a worker, but if that MediaSource blob were to be assigned to HTMLMediaElement, then it would be moved to the main loop once the user determines that the video playback is the most important to them.

**Comment by bradisbell** [comment]
@estobbart Google's MEI and autoplay policies are based on narrow assumptions about the scope of video on the web, and have actively damaged our ability to build more complex applications around client-side audio and video creation.  While Google is certainly in their right to determine what's best for their users, I don't think those policies should have much bearing in a W3C discussion.

The proposal at hand seems to have the potential to resolve some performance issues, particularly around the loading of new content.  Even if that's all that was solved (before the `srcObject` were set), it would be a significant improvement for some types of web applications.

**Comment by bmaurer** [comment]
> If video is being done in a worker, it would hint to me that video playback isn't the primary objective of the user on the site. If it is the primary objective, can't the other rendering operations be limited instead of tying up the main loop?

Wanting to run the video decoder in a worker is not incongruous with the video being the user's primary objective on the site. While the idea of trying to limit the main thread is admirable, it's often something that's easier said than done. If the user's primary objective is to view the video, running it in a dedicated thread is the best way to guarantee that we provide smooth playback and isolate the user from other things that might be going on in the main thread.

**Comment by wolenetz** [comment]
I have a proposal for incubating a solution for this issue, hopefully with eventual inclusion in MSE vNext. Please take a look:
https://github.com/wicg/media-source/blob/mse-in-workers-using-handle/mse-in-workers-using-handle-explainer.md

**Comment by jernoble** [comment]
This proposal seems unnecessarily complex; couldn't we achieve the low-startup-latency problem by adding a synchronous `open()` method, and solve the "in a worker" problem by exposing MediaSource to Worker contexts, and use the existing not-optimal-but-not-being-deprecated blob:// URL syntax as the transferrable handle to the MediaSource?

**Comment by DanielBaulig** [comment]
@wolenetz I have looked at the proposal and from an application developer perspective, this will give us the tools we need. 

As discussed before and outlined in your open questions, having the ability to create the MediaSourceHandle in the Window context and extracting the MediaSource in the Worker context would be something that we would fundamentally be interested in, but we understand that the specification and implementation complexity might not be worth it.

Other than this, we do not see any objectionable or controversial parts in the proposal, and I think we can be fully supportive.

**Comment by wolenetz** [comment]
@jernoble @DanielBaulig @bmaurer @jyavenard I have a rough Chrome prototype of MSE-in-workers now beginning to approach "working" at https://chromium-review.googlesource.com/c/chromium/src/+/1405697. It is not landable in Chromium as-is, as several items (see CL description and patch list title history) need resolution. Especially, please be aware that cross-thread teardown/MSE detachment currently hits a crash (I hope to resolve before FOMS next week, but may not have time) -- in patch set 41.

I look forward to any and all feedback, and will hopefully see many of you at FOMS/next week. Thanks!

**Comment by beaufortfrancois** [comment]
Here's some Javascript code to get a grasp of latest proposal:

```javascript
// main.js

const worker = new Worker('worker.js');
worker.postMessage('pleaseCreateMediaSource');
worker.onmessage = event => {
  document.querySelector('video').src = event.data.blobUrl;
};

// worker.js

self.onmessage = event => {
  if (event.data == 'pleaseCreateMediaSource') {
    const mediaSource = new MediaSource();
    mediaSource.open();
    self.postMessage({ blobUrl: URL.createObjectURL(mediaSource) });

    mediaSource.onsourceopen(_ => {
      // TODO: Add source buffer, fetch video chunks, ... 
    });
  }
};

**Comment by MattiasBuelens** [comment]
Small nit-pick:
> ```js
> mediaSource.onsourceopen(_ => {
>   // TODO: Add source buffer, fetch video chunks, ... 
> });
> ```
I think that's supposed to be:
```js
mediaSource.onsourceopen = _ => {
  // TODO: Add source buffer, fetch video chunks, ... 
});
```

**Comment by beaufortfrancois** [comment]
Thank you @MattiasBuelens!

**Comment by jonas-db** [comment]
MediaSource is still unavailable in a webworker context (chrome 81). Any updates?

**Comment by wolenetz** [comment]
@https://github.com/w3c/media-source/issues/175#issuecomment-630383367, thank you for your interest. After initial internal design reviews for the (new) Chromium prototype, I'm continuing work and hope to be able to announce an experimental Chrome Origin Trial for MSE-in-Workers in the short term.

Today, I've updated the original [WICG Explainer](https://github.com/wicg/media-source/blob/mse-in-workers-using-handle/mse-in-workers-using-handle-explainer.md) to reflect the simplified approach being taken in the prototype.

**Comment by tjenkinson** [comment]
Looks good. How do you access the current time in the worker? Post messaging time updates from the media element?

**Comment by wolenetz** [comment]
@https://github.com/w3c/media-source/issues/175#issuecomment-640099375 - yes, to maintain the simplicity and interoperability of browser implementations of this feature, app will need to use normal communication such as postMessage to internally communicate the state of MSE to their main thread components and the state of the media element to their worker thread components. Internally, some of the MSE algorithms require some of this on some implementations, but they will handle that as necessary in those implementations without requiring all implementations expose some new field on the object model. For instance, in Chromium, MSE's coded frame eviction algorithm relies upon internal state, including currentTime, to more conservatively keep existing buffered media around currentTime and the last appended media. However, the worker MSE will not expose that notion of currentTime to the worker app; the worker app will need to get that info from the media element on the main thread and postMessage it.

**Comment by wolenetz** [comment]
Feature discovery of whether or not an implementation supports constructing a MediaSource in a dedicated worker might need exposure of an attribute on the MediaSource interface on at least the main Window context, so API users can feature detect in advance of spinning up a dedicated worker. Credit to @joeyparrish for pointing this out to me.

**Comment by wolenetz** [comment]
@https://github.com/w3c/media-source/issues/175#issuecomment-695038710, longer term, if we end up also adding srcObject or other approaches for attaching cross-context, we could do feature detection on the object type itself perhaps. For example, can main context construct a MediaSourceHandle? If so, MSE-in-workers should be supported. Or alternatively, does main thread's MediaSource allow transferability? If so, then MSE-in-workers should be supported. The routes we can take in future around srcObject/etc are flexible and not settled at all yet.

In short term, experimental Chromium prototype that is nearing readiness right now for experimentation, we use the objectURL approach described in the updated WICG explainer, which by itself, doesn't let the main thread know whether or not a dedicated worker supports creation of a MediaSource, hence the feature discoverability might need an attribute or some other way so API users don't have to first start a dedicated worker to determine if MediaSource can be constructed there.

**Comment by wolenetz** [comment]
@https://github.com/w3c/media-source/issues/175#issuecomment-695040660 - as of Chromium 88.0.4296.0, experimental MSE-in-Workers is available **and functioning** behind experimental flag (--enable-experimental-web-platform-features). Today, I'm landing a fix for a rare Chromium crash made less rare by this feature (https://chromium-review.googlesource.com/c/chromium/src/+/2486074).

I'm also working on a way of exposing to the main thread ability for apps to proactively detect if MSE-in-Workers is supported. How does a (static, exposed wherever MediaSource is exposed) boolean-valued, always true if it exists, attribute named: "canConstructInDedicatedWorker" on the MediaSource interface sound to folks? Alternatively, are there examples in the web-platform where collections of values are exposed to enable feature-detection of a variety of features?

**Comment by joeyparrish** [comment]
`canConstructInDedicatedWorker` sounds good to me, but I'm also not terribly picky about the exact name of it.

I'm not sure what the best practice is for feature detection broadly in web API design.

In other APIs, we have done feature detection by checking for the existence of a class or the existence of a method on a prototype.  In some rare cases, we use the number of arguments to a constructor to determine which version of an API was implemented (_so many_ `TextTrackCue` variations on old devices).

**Comment by wolenetz** [comment]
Ok, I'll proceed with `canConstructInDedicatedWorker` in the experimental implementation in Chromium. I'll update here once that ability has landed (on top of the already-landed experimental feature in 88.0.4296.0 already).

**Comment by wolenetz** [comment]
As of Chromium 88.0.4300.0, experimental MSE-in-Workers is available and functioning behind experimental flag (--enable-experimental-web-platform-features), and includes support for proactive feature detection from the main thread via `MediaSource.canConStructInDedicatedWorker` as described in https://github.com/w3c/media-source/issues/175#issuecomment-712516853.

I'm putting together more details to share soon regarding how to use the feature, areas of the feature which could use further definition, and a very simple toy demo app.

**Comment by wolenetz** [comment]
Demo:
* [Avoiding main thread contention and associated "buffering jank"](https://tinyurl.com/mse-in-workers-demo)
* [Github repo for demo](https://github.com/wolenetz/mse-in-workers-demo/)

Please try it out (in Chromium 88.0.4300.0 or greater, with chrome://flags/#experimental-web-platform-features enabled.)

**Comment by johnBartos** [comment]
Hey Matt, great work! I'm wondering how to best handle this new error: `DOMException: Failed to read the 'buffered' property from 'SourceBuffer': Worker MediaSource attachment is closing`. I noticed that the readyState is still "open" in this case. Is there a recommended way to handle this new error?

**Comment by wolenetz** [comment]
@https://github.com/w3c/media-source/issues/175#issuecomment-742132101 -

Hi John. Thank you for trying out this feature. This new error condition that you're encountering is one of the three scenarios on which we have requested feedback; specifically, it is the second bulleted question in https://github.com/w3c/media-source/issues/175#issuecomment-721395481.

When handling this particular "Worker MediaSource attachment is closing" InvalidStateException scenario when readyState appears valid for the operation, the worker portion of the app that is exercising the MSE API should behave as if either:

- the main thread portion of the app has explicitly detached the MediaSource, initiating MSE closure that will happen soon in the worker,
- the main thread media element has experience a fatal error, initiating MSE closure that will happen soon in the worker, or
- the main thread media element's context is destructed, initiating MSE closure that will happen soon in the worker.

If it would help to have a distinct error (not an InvalidStateError DOMException) to help identify this scenario to the worker MSE API user, what kind of error would you prefer?
Alternatively, or in addition to a distinct error, would a new readyState, such as "closing" be appropriate? I did not include such in the experimental implementation because such state would not be guaranteed consistent across multiple synchronous reads of the attribute by the worker MSE API user (the transition to "closing" is something that could happen at any time - except within carefully protected mutex regions within the respective MSE call handling in the worker).

I look forward to your responses.

**Comment by johnBartos** [comment]
@wolenetz Thanks for the detailed response. 

>the main thread portion of the app has explicitly detached the MediaSource, initiating MSE closure that will happen soon in the worker,

Yeah this is what happened, I detached the MediaSource on error, but the worker thread continued to act upon the MediaSource. At the time of video error I send a message to stop the worker MediaSource instance, but since that's async it's not guaranteed to work.

> Alternatively, or in addition to a distinct error, would a new readyState, such as "closing" be appropriate? 

This would be my preferred choice because I think the implementation would be the simplest. Before applying any SourceBuffer operation, I could just check this flag. But if it's not consistent across reads then I'd still want to wrap operations in a try/catch, and at that point there's not much value in the flag.

> If it would help to have a distinct error (not an InvalidStateError DOMException) to help identify this scenario to the worker MSE API user, what kind of error would you prefer?

Right now I think the error is fine, but I haven't yet gotten around to detailed error handling. I don't think it'd help in this case.

I believe the correct way to do to this is to have the worker thread tell the main thread when to detach the source. This way the worker can clean up any pending operations before the detach. With this approach, is there a risk of getting errors from the MediaSource if an operation is applied to a media element in the errored state? For example, say I call `enqueue` after a decode error has been triggered.

**Comment by johnBartos** [comment]
> The reasoning for this is that the MediaSource object is known to be closing soon by the implementation, and the implementation is already destructing the underlying demuxer and player, so proceeding without giving error would be worse in this case. 

This won't always be true. Third party developers using a player SDK may bypass the proper APIs and remove the video source directly. Player vendors will therefore need to gracefully handle cases when the MediaSource is detached without advanced knowledge.

Even with a try/catch around the new exception, I'm noticing that the page still crashes after the error is handled. Is this expected in Canary?

**Comment by wolenetz** [comment]
> > Alternatively, or in addition to a distinct error, would a new readyState, such as "closing" be appropriate?
> 
> This would be my preferred choice because I think the implementation would be the simplest. Before applying any SourceBuffer operation, I could just check this flag. But if it's not consistent across reads then I'd still want to wrap operations in a try/catch, and at that point there's not much value in the flag.

It would not be consistent across reads (otherwise, the main thread would need to block itself until the attached worker thread reaches a stable state, before the main thread could continue and destruct the media player, etc.), so would we need to have a new 'closing' readyState then if try/catch is necessary already? In worst case, we might need to actually do that main thread blocking/delayed shutdown of media player if the approach to having a "closing at any moment" model in the experimental implementation is too onerous.

> > If it would help to have a distinct error (not an InvalidStateError DOMException) to help identify this scenario to the worker MSE API user, what kind of error would you prefer?
> 
> Right now I think the error is fine, but I haven't yet gotten around to detailed error handling. I don't think it'd help in this case.
> 
> I believe the correct way to do to this is to have the worker thread tell the main thread when to detach the source. This way the worker can clean up any pending operations before the detach. With this approach, is there a risk of getting errors from the MediaSource if an operation is applied to a media element in the errored state? For example, say I call `enqueue` after a decode error has been triggered.

Certainly, if the app could first prepare its worker thread for imminent main-thread detachment, and control the actual detachment sequence in the common case, this would help the app understand the attachment state better. However, there are also non-rare scenarios (such as decode error in the pipeline, or user closing the tab) whereby the main thread might begin detachment without being told to do that explicitly by the app.

I personally think a combination of a 'closing' readyState (to help apps confirm the reason for exception) and InvalidStateError Exception on operations on the worker MediaSource taken by the app when there is a pending detachment of the MediaSource, would be a good way forwards.

> > The reasoning for this is that the MediaSource object is known to be closing soon by the implementation, and the implementation is already destructing the underlying demuxer and player, so proceeding without giving error would be worse in this case.
> 
> This won't always be true. Third party developers using a player SDK may bypass the proper APIs and remove the video source directly. Player vendors will therefore need to gracefully handle cases when the MediaSource is detached without advanced knowledge.

Would the proposed 'closing' readyState and InvalidStateError Exception satisfy this concern?

> Even with a try/catch around the new exception, I'm noticing that the page still crashes after the error is handled. Is this expected in Canary?

Crashing is not expected in Canary -- unless there is an attempt to access the audio/video/textTracks() of a worker-owned SourceBuffer (media tracks in MSE/HTMLMediaElement is also experimental, and incomplete in the MSE-in-Workers experimental implementation currently). Do you have any indications of what specifically is causing the crash? If you share a crash report ID with me, from your repro's chrome://crashes (direct email wolenetz@chromium.org), then I can investigate further.

**Comment by johnBartos** [comment]
> Do you have any indications of what specifically is causing the crash?
Yeah, if you detach the MediaSource before closing it the tab will crash. Unfortunately it doesn't seem that I can enable non-local crash reporting in Chrome? It only gives me the local ID in chrome://crashes, and the toggle to enable crash reporting is unclickable for me.

My repro steps are simple though - start a MSE in workers instance and then call `$('video.src') = ''`

![image](https://user-images.githubusercontent.com/5042276/102665067-1a578880-4152-11eb-8179-f8035b8ab2b8.png)

> I personally think a combination of a 'closing' readyState (to help apps confirm the reason for exception) and InvalidStateError Exception on operations on the worker MediaSource taken by the app when there is a pending detachment of the MediaSource, would be a good way forwards.

I think this would be good. Restructuring our app to ensure MediaSource deletion wasn't onerous in my opinion, and having the `closing` state would help us handle the error correctly. I believe it addresses our concern as long as catching the error does not crash the tab 😁.

To be clear, would this `closing` state would be reliable only after the exception is thrown? Meaning that developers should not use it to prevent errors, and that the recommendation for app devs is to ensure that interactions with the MediaSource are properly wrapped in a try/catch.

**Comment by wolenetz** [comment]
> My repro steps are simple though - start a MSE in workers instance and then call `$('video.src') = ''`

@johnBartos - I could not get a crash to repro in my trunk build locally (using `$(video).src=''`, to get it to parse...). Could you perhaps share a minimized small repro page+worker js with me (privately would be fine)?

**Comment by johnBartos** [comment]
@wolenetz Sorry for the delay, will do soon.

Another question - when MSEIW is released, will it be supported on every device which is on the supported Chrome version? I want to track its usage when we roll it out and I'm wondering if I can just go by the Chrome version.

**Comment by wolenetz** [comment]
@johnBartos - I think we discussed this on slack, but for all here, MSEIW will likely first go through origin trials in Chrome; potential API users will be able to sign up for experimental support for a period of releases to assist stabilization and eventual shipping. Unless plans change, it's likely the experimental feature will be in the origin trial across the Chrome products involved in the releases. Throughout, the proactive ability for apps to be able to detect if the feature exists and is available is expected to be available at runtime via `MediaSource.canConstructInDedicatedWorker` as described, above.

@johnBartos - regarding the crash you had reported, and I had been unable to repro, is it still occurring? Do you have more details on how I could possibly repro it?

**Comment by johnBartos** [comment]
@wolenetz I'll get you simple steps (created a ticket so I won't forget). It might be an issue with my machine since I'm getting other dev tools crashes.

**Comment by johnBartos** [comment]
Ok I have steps. I used your demo but modified it to append in 1-byte increments to give me plenty of time to repro the crash:

- Start the demo with append 1-byte increments
- In the console, set the video src of the worker element to `''`
- Observe the following error:
![image](https://user-images.githubusercontent.com/5042276/112532900-c0cdae00-8d7f-11eb-8f80-ceff3bce95da.png)

Environment:
- Chrome 89.0.4389.90 (Official Build) (x86_64)

I think the key is to detatch the src and then have the MediaSource do an append. I can upload a fork with my modification if you'd like.

**Comment by wolenetz** [comment]
@johnBartos I'm trying to reproduce on trunk Chrome with a new radio button for 1-byte appends in a local version. So far, I haven't been able to get the tab to crash. Have you tried on recent Canary (maybe something unrelated, say in devtools, is causing the crash)?

**Comment by johnBartos** [comment]
Awesome, nice find Matt! I'll test out the fix once it lands

**Comment by wolenetz** [comment]
@johnBartos and any interested: I found and am fixing another issue in the experimental Chromium implementation: reading mediaSource.duration attribute from a worker context wasn't taking the right internal locks and was unsafe. https://chromium-review.googlesource.com/c/chromium/src/+/2909389 has more detail.

Also, I'm preparing a draft feature spec as part of getting MSE-in-Workers prepared for launching Chrome origin trials. Expect news on this soon.

**Comment by johnBartos** [comment]
Sounds great @wolenetz, thanks for the update! Looking forward to the origin trial

**Comment by karlt** [comment]
Having entangled objects across threads sets up an inter-thread cycle, which is a challenge for GC.  I have trouble recalling all the possible scenarios, but I guess there are at least some where operations on either object may resume playback. e.g. if playback has stalled through lack of data, then it may resume either if more data is provided through a `SourceBuffer` or if a seek is performed on the `HTMLMediaElement` to a position which already has data.
Neither thread can independently GC its object because the entangled object may want to use it again.

I assume this is a theoretically solvable problem, but complex enough that no browser solves it.  `MessagePort` entanglement is similar and the workaround is an explicit `close()` method to disentangle the objects.  Clearing the source resource on the `HTMLMediaElement` would provide a similar explicit disentengle function from the main thread.  IIUC `MediaSource#endOfStream()` is not a permanent termination and so doesn't provide a worker-thread counterpart.

Has consideration been given to mitigating inter-thread GC difficulties?

Maybe `MediaSource` and `SourceBuffer` could be garbage collected if their events depend only actions on that thread, but the "sourceclose" event depends on `HTMLMediaElement` state, which may change in response to actions on its thread.

**Comment by wolenetz** [comment]
@https://github.com/w3c/media-source/issues/175#issuecomment-896355007 you raise good points that I have encountered when designing and implementing the current experimental prototype of MSE-in-Workers in Chromium.

For same-thread attachments, GC is prevented by any of:

- an active reference to the attached media element, MediaSource, or any of its SourceBuffers (e.g. from script or transitively from other objects back to a GC roots),
- the attached media element having some pending activity (e.g. seeking or an event is pending dispatch),
- or the attached MediaSource or its SourceBuffers having some pending activity (e.g. updating or an event is pending dispatch)
- or any of the objects that may not be traceable from a GC root but have pending activity themselves and have some reference, perhaps transitively, to the attached media element, MediaSource, or its SourceBuffers.

For cross-thread attachments (e.g., for MSE-in-Workers), Blink doesn't currently have (nor does it appear to be planned from what I recall) a notion of a "CrossThreadMember" that would automate the rules, above, similarly. So, yes, cross-thread attachments need explicit teardown by detachment from the media element to break the strong reference held by the internal CrossThreadMediaSourceAttachment GC-able object, which holds the references to the attached media element and MediaSource CrossThreadMembers (which are treated themselves as GC roots on the thread owning the referenced object).

Some way of finding inactive cross-thread attachment cycles and automatically detaching them to enable their GC is not included in either the current draft specification for the feature, nor in the experimental Chromium implementation.

**Comment by wolenetz** [comment]
The experimental implementation in Chromium recently has taken a change to mitigate potential high-resolution timer security issues, so the communication of internal buffered, seekable MSE state and internal element error and recent media element time state across contexts is no longer optimized but instead has latency similar to the unoptimized postMessage() option described in the spec draft.

While this change hopefully shouldn't impact function or performance of the API significantly, please be aware that it might affect your use case.

Also, I'm working on enabling the Chromium origin trial for this feature right now and expect to be able to share more details very soon.

**Comment by wolenetz** [comment]
Thanks to @johnBartos for discovering the intentional crash in the Chromium experimental implementation when the experimental AudioVideoTracks feature usage is attempted on a worker SourceBuffer. I've [updated](https://github.com/wolenetz/mse-in-workers-demo/commit/b03366da11b36f17655a071947e562c9d94e6835) the [demo's instructions](https://tinyurl.com/mse-in-workers-demo) for how to more narrowly enable just the MediaSourceInWorkers experimental feature until both the spec issue #280 is resolved and the implementation is updated.

**Comment by wolenetz** [comment]
The Chromium experimental implementation of MSE-in-Workers is currently in origin trials (as of M95).

**Comment by sterpe** [comment]
@wolenetz What is timeline (or current thoughts) for when MSE-in-Workers would be turned on in Chrome by default?

We (Facebook) are interested in starting some active development around this in 2022.

**Comment by wolenetz** [comment]
@https://github.com/w3c/media-source/issues/175#issuecomment-932483258 The current plan is to have the origin trial run from M95 through M99, and if data and feedback show that it is ready to launch, then it could be launched around the M99-M100 milestone (Q1 2022). I highly encourage you to engage in the origin trial as early as possible to help ensure both an ergonomic API shape and a stable implementation.

**Comment by wolenetz** [comment]
Note that using srcObject for MSE-in-Workers attachments is a change to the feature that the media wg arrived at last September [1] (following the previous objectURL-based attachment design getting more recent feedback [2]), and I'm working on getting that into both the feature spec and Chromium's experimental implementation currently.

[1] https://www.w3.org/2021/09/14-mediawg-minutes.html
[2] https://github.com/mozilla/standards-positions/issues/547

**Comment by wolenetz** [comment]
With respect to switching to using srcObject-based attachment for MSE-in-Workers, I'm putting together a spec draft now. I've discussed approaches previously during the initial explainer [1], on the mozilla request for position [2], and directly with one of the current Chrome origin trial participants.

There's likely some flexibility in how that specification and implementation lands fully:

1. Should the transferable object be only constructible on a worker context? Or can it be constructed on the main context, attached before even a worker is spun up, and eventually transferred to a worker for use there in resolving the underlying MediaSource? (I prefer the former for this initial approach and for simplicity. The latter approach involves questions like, what should the state of the media element be if the worker doesn't yet have the underlying MediaSource. It is not precluded from being explored in the future though.)

2.  Must we also (as part of this origin trial) enable main-thread MSE attachment using srcObject? (I prefer this be out-of-scope of the MSE-in-Worker OT, for simplicity and priority of getting MSE-in-Worker impl(s) updated and shipped.) Likewise, it is not precluded from being explored in the future.

I have received some private support for preferred options in 1 & 2 from a current Chrome origin trial participant, those preferences being:
1: worker constructs the handle, transfers it to main, main uses the handle to attach the worker MediaSource instance via setting the element's srcObject to be the handle object, and
2: main-thread srcObject attachment of some handle object or even just the MediaSource for MSE isn't required (for at least usage in MSE-in-Workers work). Existing objectURL attachment is working already.

@jernoble (and also anyone interested in this feature), would this approach to enabling srcObject usage for MSE-in-Workers attachment, as a replacement for using objectURLs for those attachments, have your support? If so, I'll get a spec draft and implementation update readied ASAP.

**Comment by karlt** [comment]
> * Should the transferable object be only constructible on a worker context? Or can it be constructed on the main context, attached before even a worker is spun up, and eventually transferred to a worker for use there in resolving the underlying MediaSource? (I prefer the former for this initial approach and for simplicity. The latter approach involves questions like, what should the state of the media element be if the worker doesn't yet have the underlying MediaSource. It is not precluded from being explored in the future though.)

The two options presented here are quite different approaches, requiring someone different APIs, so I'd be inclined to choose one or the other.  The reasons you've given for the former approach sound good to me.

Perhaps, for consistency, the object could be constructed in Window context (as well as in Worker context) from a MediaSource in the same Window context for attachment to video element in the same or another Window context.  But I don't feel this should block progress for Worker.  It would not be precluded in the future.

> * Must we also (as part of this origin trial) enable main-thread MSE attachment using srcObject? (I prefer this be out-of-scope of the MSE-in-Worker OT, for simplicity and priority of getting MSE-in-Worker impl(s) updated and shipped.) Likewise, it is not precluded from being explored in the future.

Similarly direct attachment of Window-context MediaSource to srcObject seems orthogonal to transferable object attachment, and need not block progress for Worker.

**Comment by wolenetz** [comment]
@jernoble  - friendly ping to hopefully get your input on https://github.com/w3c/media-source/issues/175#issuecomment-1042395368

**Comment by wolenetz** [comment]
Chrome's MSE-in-Workers origin trial end milestone has now been extended to M103 to accommodate time for the srcObject attachment mechanism change to this feature. I'll update here when the change to add srcObject for attachment has landed in Chromium, which will then shortly be followed by the removal of the objectURL-based attachment mechanism (ideally in distinct milestones to assist trial participants' implementations picking amongst the mechanisms).

**Comment by wolenetz** [comment]
During Media WG meeting today [1], @jernoble affirmed the preferred approach in https://github.com/w3c/media-source/issues/175#issuecomment-1042395368 towards which I have optimistically been drafting the spec update. I'll continue my work on that.

[1] https://www.w3.org/2022/03/08-mediawg-minutes.html

**Comment by wolenetz** [comment]
Spec draft for the proposed srcObject way of attaching MSE-in-Worker (and disallowing MSE-in-Worker attachment via object URL) is uploaded and PR is pending review by @mwatson2 currently. See #305

**Comment by wolenetz** [comment]
Spec is updated. Chromium implementation and WPT have landed for it (as of 105.0.5180.0).
Today, I:
* updated the [demo](https://tinyurl.com/mse-in-workers-demo) to require handle+srcObject (instead of legacy object URL) for worker MediaSource attachment, and
* updated the [original WICG explainer](https://github.com/wicg/media-source/blob/mse-in-workers-using-handle/mse-in-workers-using-handle-explainer.md) similarly to match the current specification and Chromium experimental implementation.

**Comment by wolenetz** [comment]
I have also sent for review by blink-dev (part of Chrome review launch process) intents to:
* ship MSE-in-Workers (using MediaSourceHandle as specified currently) in Chrome 105 (pending their approvals currently): https://groups.google.com/a/chromium.org/g/blink-dev/c/FRY3F1v6Two
* extend the origin-trial in Chrome through 104 (using the legacy worker MediaSource object URL attachment mechanism, but otherwise unchanged internally once attached versus the current MediaSourceHandle API shape): https://groups.google.com/a/chromium.org/g/blink-dev/c/WETrrUYLrTM

I'll update here when one or both of those intents are approved.

**Comment by wolenetz** [comment]
> I'll update here when one or both of those intents are approved.

Both of the blink intents have been approved.

- MSE-in-Workers (using the spec's MediaSourceHandle and srcObject for attachment) is enabled in Chrome 106 by default (beginning with 106.0.5214.0) and I expect for it to be available soon in Chrome 105 once further approvals are received.

- MSE-in-Workers origin trial (using the earlier experimental API for this, where attachment of worker MediaSource is achievable by getting an object URL for it over to the main thread and assigning it to the media element's src attribute) is now extended to Chrome 104, ending August 30, 2022 (https://developer.chrome.com/origintrials/#/view_trial/3847199981681246209).

Note that the latter (origin trial) version of the API is NOT how the shipping version (and spec'ed version) works, so origin trial participants will need to expect to update their API usage accordingly based on which version of Chrome (OT through 104, shipping version beginning soon in 105) is being used. The presence of a 'handle' getter on the MediaSource prototype indicates the expected API that is shipping soon in 105. Likewise, the presence of a 'MediaSourceHandle' on the context indicates the same. Regardless, if the static method 'MediaSource.canConstructInDedicatedWorker' is missing or exists but returns false, then neither version of the API is available.

Some further points about this API transition from origin trial to the shipping version (again, expected to become enabled in Chrome 105 soon - I'll update here again when that happens):

- Though you can still obtain an "object URL" for a dedicated worker MediaSource, it'll fail with media element error if you try to use it to attach to the media element.
- Instead, obtain a "MediaSourceHandle" from the dedicated worker MediaSource (using a new readonly [SameObject] getter: "handle"), and transfer it (not just serialize it) to your main thread. Then assign it to the media element's srcObject attribute. Example of change in worker API usage:
  - const ms = new MediaSource(); const url = URL.createObjectURL(ms); postMessage(url); // OLD API shape (through 104)
  - const ms = new MediaSource(); const handle = ms.handle; postMessage(handle, [handle]);  // New API shape (shipping in 105 soon)
- Important points about the new API shape.
  - A MediaSourceHandle instance is usable at most once for successfully attaching to a media element. A Worker MediaSource instance has precisely one MediaSourceHandle instance. Therefore, for multiple MediaSource attachments, you'll need to obtain new worker MediaSource instances and transfer their handles, for each attachment.
  - No worker MediaSource object URL revocation is necessary, since worker MediaSource now attaches via a MediaSourceHandle object and the implementation knows how to GC these objects when there remain no references to them and their attachment has been closed (for example, by assigning "null" to the media element's srcObject attribute.)
- There are other fine points. I recommend reviewing the current MSE spec working draft for edge cases not outlined above: https://www.w3.org/TR/media-source-2/

**Comment by wolenetz** [comment]
> ...expected to become enabled in Chrome 105 soon - I'll update here again when that happens):

* If using Chrome 106.0.5214.0 or greater, the feature is enabled by default.
* If using Chrome 105.0.5195.18 or greater in the 105 milestone, the feature is
  enabled by default.

**Comment by Aholicknight** [comment]
<details>
  <summary>Images</summary>

![image](https://user-images.githubusercontent.com/7843719/187298709-9af28104-0fcb-49f8-93f1-29d49306227b.png)

![image](https://user-images.githubusercontent.com/7843719/187298767-07848fd3-c8fb-4be7-929e-4b7424853997.png)
</details>

I am getting the same error with the latest version of Microsoft Edge Beta v105.0.1343.17 is there any way to fix this? (Please check images in the summary)

Thank you!

**Comment by dalecurtis** [comment]
Actually a more elegant solution may be to just tag "handle" as [Exposed=DedicatedWorker]" since it just throws an exception on the main thread anyways. We already have canConstructInDedicatedWorker for feature detection.

**Comment by Aholicknight** [comment]
> This is tracked by [bugs.chromium.org/p/chromium/issues/detail?id=1358542](https://bugs.chromium.org/p/chromium/issues/detail?id=1358542) it looks like an old version of video.js is trying to mirror all properties on the MediaSource object, so throwing an error upon access to handle is exploding in the main thread.
> 
> I think we'll want to return nullptr instead of throwing an exception when Handle is accessed from the main thread.

@dalecurtis so this is the sites problem and not mine?

**Comment by wolenetz** [comment]
I'm finally back from vacation and illness recovery after it. I'll be looking at the MSE-in-Workers issue ASAP as I catch up.

**Comment by wolenetz** [comment]
For those interested in contributing to the discussion around how to update the spec for the MediaSource.handle getter, please see #316.

Also, I've updated the [MSE in Workers demo](https://tinyurl.com/mse-in-workers-demo) today:
1. It no longer checks for the "handle" getter on the Window/Main context MediaSource object, since "handle" will only be needed by the demo on the dedicated worker context (https://github.com/wolenetz/mse-in-workers-demo/commit/056a416649a35a10e8a563e3189cfe4ef5f06f28).
2. Updated the "known-working" Chrome versions listed in the demo and in its README (https://github.com/wolenetz/mse-in-workers-demo/commit/9685c9470abe089f35d90e97e6dd05415e2e0083).

**Comment by wolenetz** [comment]
@https://github.com/w3c/media-source/issues/175#issuecomment-1242390314 thanks for your input. We're currently considering two options: Change from exception to null value for the 'handle' attribute when accessed on a main thread MediaSource, or change visibility of that handle getter to only be visible on the dedicated worker context (for now at least). I presume we'll disallow exception even in the latter case.

Either way, it looks like, on main thread (or on worker thread), `MediaSource.canConstructInDedicatedWorker === true` will remain the overall feature detection for MSE-in-Workers, and, when it's true, both dedicated worker and main thread contexts should have `typeof MediaSourceHandle === 'function'` on conforming implementations.

**Comment by wolenetz** [comment]
From discussion during Media Working Group meeting at TPAC 2022 today, both @jernoble and @dontcallmedom also supported the visibility restriction option for the handle getter. I presented both options (null with custom spec replacement for [SameObject], or just exposed=dedicatedworker -- neither can throw exception).

I'll update (or just abandon/replace) the PR #316 to instead do visibility restriction of that getter, without exception throwing.

**Comment by wolenetz** [comment]
In Chrome, the feature has been fixed to behave per #317 and builds of Chrome >= 108.0.5334.0 should include it enabled by default. See also the Chromium feature tracking [issue 878133](https://bugs.chromium.org/p/chromium/issues/detail?id=878133).
